source/common/rsa.c

```diff
@@ -6,22 +6,32 @@
 #include <nds/rsa.h>
 #include <nds/system.h>
 
-int swiRSAInitHeapTWL(swiRSAHeapContext_t *ctx, void *heapStart, size_t heapSize);
-int swiRSADecryptRAWTWL(swiRSAHeapContext_t *ctx, swiRSAbuffers_t *rsabuffers, size_t len);
+#include <stdlib.h>
+
+void swiRSAInitHeapTWL(swiRSAHeapContext_t *ctx, void *heapStart, size_t heapSize);
+int swiRSADecryptRAWTWL(swiRSAHeapContext_t *ctx, swiRSAbuffers_t *rsabuffers, size_t *lenDst);
 int swiRSADecryptTWL(swiRSAHeapContext_t *ctx, void *dst, const void *sig, const void *key);
 int swiRSADecryptPGPTWL(swiRSAHeapContext_t *ctx, void *dst, const void *sig, const void *key);
 
-int swiRSAInitHeap(swiRSAHeapContext_t *ctx, void *heapStart, size_t heapSize)
+swiRSAHeapContext_t createRSAHeapContext(size_t heapSize)
+{
+    swiRSAHeapContext_t rsaCtx;
+    rsaCtx.heapSize = (heapSize == 0) ? SWI_RSA_DEFAULT_HEAP_SIZE : heapSize;
+    rsaCtx.heapStart = malloc(rsaCtx.heapSize);
+    rsaCtx.heapEnd = ((uint8_t *)rsaCtx.heapStart) + rsaCtx.heapSize;
+    return rsaCtx;
```

Comment:
I think that what asie meant was to create a function that is essentially:

```
int swiRSADecryptRAWTWL(swiRSAbuffers_t *rsabuffers, size_t *lenDst)
{
    swiRSAHeapContext_t rsaCtx = { 0 };
    rsaCtx.heapSize = ...
    ...

    return swiRSADecryptRAW(&rsaCtx, rsabuffers, lenDst)
}
```

Not a function that allocates the context.

Reply:
Also, remember to check if `malloc()` returns `NULL`.